### ISSUE TITLE
Multiple "load balancer has changed during request" errors

### ISSUE THREAD

**Issue by JohnnyCrazy** [hetznercloud/terraform-provider-hcloud#182]
Hi,

First of all, thanks for the provider, working fine so far 👍 

However, I noticed that working with the LB can be quite troublesome since it often fails with `load balancer has changed during request` when adding `targets` or `services`.

### services

My initial setup regarding services was like this:

```terraform
resource "hcloud_load_balancer_service" "kube_api_service" {
  load_balancer_id = hcloud_load_balancer.kube_api_load_balancer.id
  protocol         = "tcp"
  listen_port      = 6443
  destination_port = 6443
}

resource "hcloud_load_balancer_service" "http_service" {
  load_balancer_id = hcloud_load_balancer.kube_api_load_balancer.id
  protocol         = "tcp"
  listen_port      = 80
  destination_port = 80
}

resource "hcloud_load_balancer_service" "https_service" {
  load_balancer_id = hcloud_load_balancer.kube_api_load_balancer.id
  protocol         = "tcp"
  listen_port      = 443
  destination_port = 443
}
```

This setup resulted in a race condition: The first `service` was created, but the others failed with the above `load balancer has changed during request`. Fixing this was quite easy using `depends_on`, introducing a sequential chain of adding the services.

### targets

I was not able to solve the same issue for `targets`, since they depend on a dynamic array of nodes, like this:

```terraform
resource "hcloud_load_balancer_target" "kube_api_load_balancer_target" {
  count = length(hcloud_server.nodes)

  load_balancer_id = hcloud_load_balancer.kube_api_load_balancer.id
  type             = "server"
  server_id        = hcloud_server.nodes[count.index].id
  use_private_ip   = true

  depends_on = [
    hcloud_server_network.srvnetwork,
  ]
}
```

I didn't find a way to sequentially create resources when `count` is used.

So in the end I'm wondering why it doesn't work in a parallel manner. The `targets` and `services` should be independent of each other, shouldn't they?

**Comment by LKaemmerling** [comment]
Hey @JohnnyCrazy,

thank you for the reporting. We are already working on a fix for this.

**Comment by vmozgovoy** [comment]
Same trouble.

**Comment by vmozghovyi** [comment]
Hi @LKaemmerling 
I tried to fix a number of problems associated with this report. But I could not deal with the autotests, can you see what is wrong in PR #185?

**Comment by vmozghovyi** [comment]
It is my terraform manifest that is worked fine:
```
resource "hcloud_network" "network" {
  name = "k8s-net"
  ip_range = var.network_cidr
}

resource "hcloud_server" "workers" {
  count = local.workers.count
  name = "worker-${format("%02d", count.index + 1)}.${var.cluster_name}"
  image = var.cluster_image
  server_type = local.workers.server_type
  location = element(var.locations, count.index)
  ssh_keys = [hcloud_ssh_key.default.id]
  user_data = data.template_file.master-init.rendered
}

resource "hcloud_network_subnet" "workers" {
  count = local.workers.count > 0 ? 1 : 0
  network_id = hcloud_network.network.id
  type = "server"
  network_zone = "eu-central"
  ip_range = local.workers.subnet
}

resource "hcloud_server_network" "workers" {
  depends_on = [hcloud_network_subnet.workers]
  count = local.workers.count
  server_id = hcloud_server.workers[count.index].id
  network_id = hcloud_network.network.id
  ip = cidrhost(local.workers.subnet, count.index + 11)

  provisioner "remote-exec" {
    connection {
      host          = length(hcloud_server.bastions) > 0 ? self.ip : hcloud_server.masters[count.index].ipv4_address
      private_key   = local.workers.connection_ssh_private_key
      agent         = local.workers.connection_agent
      bastion_host  = local.bastion.host
      bastion_user  = local.bastion.user
      bastion_port  = local.bastion.port
      bastion_private_key = local.bastion.private_key
      timeout       = "60s"
    }
    inline = ["cloud-init status --wait"]
  }
}

resource "hcloud_load_balancer" "workers" {
  count = length(local.workers.hcloud_balancers)
  name  = "balancer-${local.workers.hcloud_balancers[count.index].name}-${format("%02d", count.index + 1)}.${var.cluster_name}"
  load_balancer_type = local.workers.hcloud_balancers[count.index].type
  location = local.workers.hcloud_balancers[count.index].location
  labels = local.workers.hcloud_balancers[count.index].labels
}

resource "hcloud_load_balancer_network" "workers" {
  depends_on = [hcloud_network_subnet.workers]
  count = length(local.workers.hcloud_balancers)
  load_balancer_id = hcloud_load_balancer.workers[count.index].id
  network_id = hcloud_network.network.id
}

resource "hcloud_load_balancer_service" "workers" {
  depends_on = [hcloud_load_balancer.workers]
  count = length(local.workers_hcloud_balancers_services)
  load_balancer_id = local.workers_hcloud_balancers_services[count.index].load_balancer_id
  protocol = local.workers_hcloud_balancers_services[count.index].protocol
  listen_port = local.workers_hcloud_balancers_services[count.index].listen_port
  destination_port = local.workers_hcloud_balancers_services[count.index].destination_port
  proxyprotocol = local.workers_hcloud_balancers_services[count.index].proxyprotocol
  dynamic "health_check" {
    for_each = local.workers_hcloud_balancers_services[count.index].health_checks
    content {
      protocol = health_check.value.protocol
      port = health_check.value.port
      interval = health_check.value.interval
      timeout = health_check.value.timeout
    }
  }
}

resource "hcloud_load_balancer_target" "workers" {
  depends_on       = [hcloud_server_network.workers]
  count            = length(local.workers_hcloud_balancers_targets)
  type             = "server"
  load_balancer_id = local.workers_hcloud_balancers_targets[count.index].load_balancer_id
  server_id        = local.workers_hcloud_balancers_targets[count.index].server_id
  use_private_ip   = true
}

locals {
  workers = merge(
  {
    count = 1
    server_type = "cx41"
    init_template = "templates/cluster-init.tpl"
    subnet = "10.0.3.0/24"
    connection_ssh_private_key = ""
    connection_agent = true
    labels = {
      node_group = "workers"
    }
    hcloud_balancers = [
      {
        name = "workers"
        type = "lb11"
        location = "fsn1"
        labels = {
          node_group = "workers"
        }
        services = [
          {
            protocol = "tcp"
            listen_port = 80
            destination_port = 32080
            proxyprotocol = true
            health_checks = [
              {
                protocol = "tcp"
                port = "32080"
                interval = 15
                timeout = 10
              }
            ]
          },
          {
            protocol = "tcp"
            listen_port = 443
            destination_port = 32443
            proxyprotocol = true
            health_checks = [
              {
                protocol = "tcp"
                port = "32443"
                interval = 15
                timeout = 10
              }
            ]
          }
        ]
      }
    ]
  },
  var.workers
  )
  workers_hcloud_balancers_services = flatten([
    for key, balancer in local.workers.hcloud_balancers :
    [
      for service in balancer.services :
      {
        load_balancer_id = matchkeys(hcloud_load_balancer.workers.*.id, hcloud_load_balancer.workers.*.name, list("balancer-${balancer.name}-${format("%02d", key + 1)}.${var.cluster_name}"))[0]
        protocol = service.protocol
        listen_port = service.listen_port
        destination_port = service.destination_port
        proxyprotocol = service.proxyprotocol
        health_checks = service.health_checks
      }
    ]
  ])
  workers_hcloud_balancers_targets = flatten([
  for balancer in hcloud_load_balancer.workers :
  [
  for server in hcloud_server.workers :
  {
    load_balancer_id = balancer.id
    server_id = server.id
  }
  ]
  ])
}

```

**Comment by fhofherr** [comment]
Hi @vmozgovoy,

first of all: thank you very much for your pull request :slightly_smiling_face: 

> I tried to fix a number of problems associated with this report. But I could not deal with the autotests, can you see what is wrong in PR #185?

The test failure seems to be related to #187. I'm working on a fix for it.

The retry issue should be fixed by `hcloud-go` version 1.18.2. I took the liberty of extracting your fix for the missing default listen port into PR #186.

**Comment by vmozghovyi** [comment]
Hi, @fhofherr 
Thank you for working on these improvements. The balancer as a service especially through terraform will be a great feature from hetzner :).

**Comment by vmozghovyi** [comment]
> Hi @vmozgovoy,
> 
> first of all: thank you very much for your pull request 
> 
> > I tried to fix a number of problems associated with this report. But I could not deal with the autotests, can you see what is wrong in PR #185?
> 
> The test failure seems to be related to #187. I'm working on a fix for it.
> 
> The retry issue should be fixed by `hcloud-go` version 1.18.2. I took the liberty of extracting your fix for the missing default listen port into PR #186.

So, if CodeConflict issue has been fixed, maybe it makes sense to remove code blocks like: 
```
	if hcloud.IsError(err, hcloud.ErrorCodeConflict) || hcloud.IsError(err, hcloud.ErrorCodeLocked) {
		// Retry after a short delay
		time.Sleep(time.Second)
		action, _, err = client.LoadBalancer.RemoveServerTarget(ctx, lb, tgt.Server.Server)
	}
```
They are found in many places.

**Comment by fhofherr** [comment]
> 
> So, if CodeConflict issue has been fixed, maybe it makes sense to remove code blocks like:
> 
> ```
> 	if hcloud.IsError(err, hcloud.ErrorCodeConflict) || hcloud.IsError(err, hcloud.ErrorCodeLocked) {
> 		// Retry after a short delay
> 		time.Sleep(time.Second)
> 		action, _, err = client.LoadBalancer.RemoveServerTarget(ctx, lb, tgt.Server.Server)
> 	}
> ```
> 
> They are found in many places.

We are currently working on various code improvements.

**Comment by LKaemmerling** [comment]
Should be fixed as of v1.19.0.
